agent/structs/prepared_query.go

```diff
@@ -227,6 +233,35 @@ func (q *PreparedQueryExecuteRequest) RequestDatacenter() string {
 	return q.Datacenter
 }
 
+// CacheInfo implements cache.Request allowing requests to be cached on agent.
+func (q *PreparedQueryExecuteRequest) CacheInfo() cache.RequestInfo {
+	info := cache.RequestInfo{
+		Token:          q.Token,
+		Datacenter:     q.Datacenter,
+		MinIndex:       q.MinQueryIndex,
+		Timeout:        q.MaxQueryTime,
+		MaxAge:         q.MaxAge,
+		MustRevalidate: q.MustRevalidate,
+	}
+
+	// To calculate the cache key we hash over all the fields that affect the
+	// output other than Datacenter and Token which are dealt with in the cache
+	// framework already.
+	v, err := hashstructure.Hash([]interface{}{
+		q.QueryIDOrName,
+		q.Limit,
+		q.Connect,
+	}, nil)
```

Comment:
This probably doesn't matter, but this will imply ordering on the hash value. The case this might matter in theory is if we ever _persist_ the cache (we **do not** now) to support "warm starts" of the client. 

If you want to nip this in the bud, you can tag a field and hashstructure will do the right thing. You may also want to consider just a full anonymous struct with fields set as well. 

```go
hashstructure.Hash(struct {
  Fields []interface{} `hash:"set"`
}{
  Fields: []interface{}{ ... },
})
```

Might not matter now though, probably worth thinking about both ways though.

Reply:
Intersting.

Yeah I'd assumed that changing order here would invalidate current caches, could certainly call that out.

In the case of a full anonymous struct, it seems that would not solve the problem - if we changed the order of the fields in the source it would presumably change the order in which reflection iterates them right? So changing source order in either case would invalidate current caches? At least that seems to be the case given the following which may not be what you mean:

```go
hashstructure.Hash(struct {
  		QueryIDOrName string
		Limit int
		Connect bool
}{
  		QueryIDOrName: q.QueryIDOrName,
		Limit: q.Limit,
		Connect: q.Connect,
})
```

And that someone coming later and adding a field not at the end like this woudl cause the same problem as with the plain list I have here right?

```go
hashstructure.Hash(struct {
  		QueryIDOrName string
		Limit int
                OtherThing bool
		Connect bool
}{
  		QueryIDOrName: q.QueryIDOrName,
		Limit: q.Limit,
                OtherThing: q.OtherThing,
		Connect: q.Connect,
})
```

I think for now given that it doesn't matter a comment about ordering will suffice since if we later rely on order (e.g. for persistence) we could make the change then?